java-frontend/src/main/java/org/sonar/java/se/ProgramState.java

```diff
@@ -125,37 +126,41 @@ private ProgramState(ProgramState ps, PMap<SymbolicValue, Constraint> newConstra
   }
 
   ProgramState stackValue(SymbolicValue sv) {
-    Deque<SymbolicValue> newStack = new LinkedList<>(stack);
-    newStack.push(sv);
-    return new ProgramState(this, newStack);
+    return new ProgramState(this, stack.push(sv));
   }
 
   ProgramState clearStack() {
-    return unstackValue(stack.size()).state;
+    return unstackValue(Integer.MAX_VALUE).state;
   }
 
   public Pop unstackValue(int nbElements) {
-    if (nbElements == 0) {
-      return new Pop(this, Collections.<SymbolicValue>emptyList());
+    if (nbElements == 0 || stack.isEmpty()) {
+      return new Pop(this, Collections.emptyList());
     }
-    Preconditions.checkArgument(stack.size() >= nbElements, nbElements);
-    Deque<SymbolicValue> newStack = new LinkedList<>(stack);
+
+    // this can perhaps be made more efficient by returning sub collection of PStack instead of copying to the list, but it's not a problem now
```

Comment:
If this is a fixme, please add the tag in front of it : sonarqube keep track of those. And avoid temporal remark (for now) which are not really appropriate in a comment IMO.